domain/agentbinary/service/simplestreamstore.go

```diff
@@ -0,0 +1,194 @@
+type ProviderForAgentBinaryFinder interface {
+	environs.BootstrapEnviron
+}
+
+// PreferredSimpleStreamsFunc is a function that returns the preferred streams
+// for the given version and stream.
+type PreferredSimpleStreamsFunc func(
+	vers *semversion.Number,
+	forceDevel bool,
+	stream string,
+) []string
+
+// AgentBinaryFilter is a function that filters agent binaries based on the
+// given parameters. It returns a list of agent binaries that match the filter
+// criteria.
+type AgentBinaryFilter func(
+	ctx context.Context,
+	ss envtools.SimplestreamsFetcher,
+	env environs.BootstrapEnviron,
+	majorVersion,
+	minorVersion int,
+	streams []string,
+	filter coretools.Filter,
+) (coretools.List, error)
+
+// GetEnvironAgentBinariesFinder returns a function that can be used to find
+// agent binaries from the simplestreams data sources.
+func (s *SimpleStreamAgentBinaryStore) GetEnvironAgentBinariesFinder() EnvironAgentBinariesFinderFunc {
+	return func(
+		ctx context.Context,
+		major,
+		minor int,
+		version semversion.Number,
+		requestedStream string,
+		filter coretools.Filter,
+	) (_ coretools.List, err error) {
+		ctx, span := trace.Start(ctx, trace.NameFromFunc())
+		defer span.End()
+
+		provider, err := s.providerForAgentBinaryFinder(ctx)
+		if errors.Is(err, coreerrors.NotSupported) {
+			return nil, errors.Errorf("getting provider for agent binary finder %q", coreerrors.NotSupported)
+		}
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		cfg := provider.Config()
+		if requestedStream == "" {
+			requestedStream = cfg.AgentStream()
+		}
+
+		streams := s.getPreferredSimpleStreams(&version, cfg.Development(), requestedStream)
+		ssFetcher := simplestreams.NewSimpleStreams(simplestreams.DefaultDataSourceFactory())
+		return s.agentBinaryFilter(ctx, ssFetcher, provider, major, minor, streams, filter)
+	}
+}
+
+func (s *SimpleStreamAgentBinaryStore) GetAgentBinary(ctx context.Context, ver coreagentbinary.Version, stream coreagentbinary.AgentStream) (io.ReadCloser, int64, string, error) {
+	finder := s.GetEnvironAgentBinariesFinder()
+	toolList, err := finder(ctx, ver.Number.Major, ver.Number.Minor, ver.Number, string(stream), coretools.Filter{
+		Number: ver.Number,
+		Arch:   ver.Arch,
+	})
+	if err != nil {
+		return nil, 0, "", errors.Errorf("getting tool list for version %q", ver.String())
+	}
+
+	if len(toolList) == 0 {
+		return nil, 0, "", errors.Errorf("getting agent binary for version %q", ver.String()).Add(domainagenterrors.NotFound)
+	} else if len(toolList) != 1 {
+		return nil, 0, "", errors.Errorf("multiple tools found for version: %q", ver.String())
+	}
+
+	tool := toolList[0]
+	toolURL := tool.URL
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, toolURL, nil)
+	if err != nil {
+		return nil, 0, "", errors.Capture(err)
+	}
+
+	resp, err := s.httpClient.Do(req)
+	if err != nil || resp == nil {
+		return nil, 0, "", errors.Errorf("error fetching tool metadata for version %q", ver.String())
+	}
```

Comment:
Don't drop the error on the floor if you're not going to log it.

```suggestion
	if err != nil {
		return nil, 0, "", errors.Errorf("error fetching tool metadata for version %q: %w", ver.String(), err)
	} else if resp == nil {
		return nil, 0, "", errors.Errorf("error fetching tool metadata for version %q", ver.String())
	}
```

Reply:
noted and refactored w thanks!